qa/workunits/rgw/test_rgw_d4n.py

```diff
@@ -127,11 +127,11 @@ def test_cache_methods(r, client, obj):
     attrs = json.loads(output.decode('latin-1'))
 
     assert((data.get(b'user.rgw.tail_tag')) == attrs.get('attrs').get('user.rgw.tail_tag').encode("latin-1") + b'\x00')
-    assert((data.get(b'user.rgw.pg_ver')) == attrs.get('attrs').get('user.rgw.pg_ver').encode("latin-1") + b'\x00\x00\x00\x00\x00\x00\x00')
+    assert((data.get(b'user.rgw.pg_ver')) == attrs.get('pg_ver').encode("latin-1") + b'\x00\x00\x00\x00\x00\x00\x00')
```

Comment:
@samarahu can you help me understand why these tests expect the redis response to have extra zero bytes at the end?

Reply:
> it is best to write the object's attributes and data into the Redis cache as is and not attempt to modify them in any way

if we were writing them to redis unmodified, then wouldn't we expect the direct comparison between `data.get()` and `attrs.get()` to succeed? these tests for extra bytes at the end makes it sound like we're storing the wrong thing in redis

cc @mattbenjamin

Reply:
The `cpp_redis` library requires a variable of type `string` to be passed to its methods, so we aren't able to directly write the original buffers. I'm not sure where the padding comes from but decoding the data results in the original buffer values in both the source code and the test suite code. However, I chose to encode the attributes into bytes rather than decode the redis data into a string because there were errors resulting from the latter. The padding does not influence the length of the buffer.

Reply:
I think the additional padding is coming from the original serialisation of the data into bufferlists.  Redis is being fed the raw buffer values converted to strings via `buffer::list::to_str()`, which copies the entire buffer length into the string including any padding bytes required by the underlying data type.  For example, `pg_ver` here is a 64-bit integer, so is always encoded to a buffer list of 8 bytes, for example a `pg_ver` of `123` will encode to `7B 00 00 00 00 00 00 00`.  This is converted to a string of length 8 by `buffer::list::to_str()`, even though the final bytes are all null.  This is sent to redis, which stores the padding explicitly, and returns it when asked.  Thus, the values sent back are padded.

Other values are having things like null terminator bytes added, as they were appended to the bufferlist directly using `.c_str()`, which includes a null terminator in the length sent.

Reply:
I understand; that certainly makes sense! However, I have observed that the length of the attribute values that are returned from Redis in the Teuthology tests do not include the padding bytes and are accurate.